lib/upgradable-multisig/multisig/utils.ak

```diff
@@ -47,3 +52,27 @@ pub fn signed_within_threshold(
 ) -> Bool {
   has_enough_signers(datum.signers, datum.threshold, extra_signatories)
 }
+
+/// Calculate a unique token name from a `TxId#Idx` and prefix. Can be combined
+/// with the `find` module to create unique token names from the first input
+/// utxo inside the transaction.
+///
+/// ```aiken
+/// values.unique_token_name(tx_id, tx_idx, cip68.prefix_333)
+/// ```
+pub fn unique_token_name(
+  txid: Hash<Blake2b_256, Transaction>,
+  idx: Int,
+  prefix: ByteArray,
+) -> AssetName {
+  let txid_hash: ByteArray = sha3_256(txid)
+  let prepend_index: ByteArray = bytearray.push(txid_hash, idx)
+  let prepend_prefix: ByteArray = bytearray.concat(prefix, prepend_index)
+  bytearray.slice(prepend_prefix, 0, 31)
+}
+
+// TODO: Find out whether token name is 28 or 32 bytes.
+pub fn multisig_token_name(own_ref: OutputReference) -> ByteArray {
```

Comment:
Why are we generating a unique token name?

Reply:
As a protection against double satisfaction. Without this prevention (or an alternative solution), authentication NFTs have a way of leaving the script address.

Although I'm not entirely sure whether this really can be an attack vector, since all these transactions are meant to be signed.